### ISSUE TITLE
Search: add 404 search recommendation block

### ISSUE THREAD

**Issue by gibrown** [Automattic/jetpack#17207]
Add a recommendation block that is displayed on 404 pages and can function as recommendations.

This could be a fork of the existing Related Posts block to start and we just need to add an option to the search API to treat the query string more as context rather than a requirement when matching.

**Comment by robfelty** [comment]
I am not sure I understand why this query would be different than just a simple search. Perhaps I don't understand the use case correctly. It seems to me like if I end up at a url like example.com/foobar, which doesn't exist, then searching for "foobar" via the normal search query would give me some relevant stuff. The only main difference I can think of is whether we might try to do some auto-correction of spelling mistakes, but I am not sure if we are doing that anywhere yet.

**Comment by gibrown** [comment]
I don't think we can treat it as an AND of all terms. Closer to a related posts query where we want to find the best thing that matches. And probably we should show X of the best things that match. Sort of why it feels similar to Fairybot to me, but I guess that has cases where you want to return nothing, or have some cut-off.

An example would be example.com/foo-bar, but the existing page is example.com/foobar. Or example.com/foobar-bear and we can recommend example.com/foobar and example.com/bear

The easiest way to treat this may just be to add a new "sort" option. "best-match" or something which will always return some results as long as everything is not filtered out. Technically though this could also be an extension of the Related Posts feature.
